sim/pkg/simulator/homogenous_stock.go

```diff
@@ -0,0 +1,75 @@
+package simulator
+
+import "fmt"
+
+type homogenousStock struct {
+	name       StockName
+	stocksKind EntityKind
+	stock      []*Entity
+}
+
+func (hs *homogenousStock) Name() StockName {
+	return hs.name
+}
+
+func (hs *homogenousStock) KindStocked() EntityKind {
+	return hs.stocksKind
+}
+
+func (hs *homogenousStock) Count() uint64 {
+	return uint64(len(hs.stock))
+}
+
+func (hs *homogenousStock) EntitiesInStock() []*Entity {
+	return hs.stock
+}
+
+func (hs *homogenousStock) Add(entity Entity) error {
+	if entity == nil {
+		return fmt.Errorf("could not add Entity, as it was nil")
+	}
+
+	if entity.Kind() != hs.KindStocked() {
+		return fmt.Errorf(
+			"stock '%s' could not stock entity '%s'; stock accepts '%s' but kind is '%s'",
+			hs.Name(),
+			entity.Name(),
+			hs.KindStocked(),
+			entity.Kind(),
+		)
+	}
+
+	hs.stock = append(hs.stock, &entity)
+	return nil
+}
+
+func (hs *homogenousStock) Remove(entity *Entity) Entity {
```

Comment:
Here you could panic if you are given a non-nil entity, since homogeneous stocks do not support removing a specific entity.

Reply:
Not relevant anymore, as we have arrayStock and mapStock which both can remove a particular entity.